src/main/java/com/google/cloud/genomics/dataflow/pipelines/CountReads.java

```diff
@@ -0,0 +1,182 @@
+import com.google.cloud.genomics.dataflow.readers.bam.Reader;
+import com.google.cloud.genomics.dataflow.utils.DataflowWorkarounds;
+import com.google.cloud.genomics.dataflow.utils.GCSOptions;
+import com.google.cloud.genomics.dataflow.utils.GenomicsDatasetOptions;
+import com.google.cloud.genomics.dataflow.utils.GenomicsOptions;
+import com.google.cloud.genomics.utils.Contig;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.cloud.genomics.utils.Paginator;
+import com.google.cloud.genomics.utils.Paginator.ShardBoundary;
+import com.google.common.base.Function;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+
+import org.codehaus.jackson.annotate.JsonIgnore;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ * Simple read counting pipeline, intended as an example for reading data from 
+ * APIs OR BAM files and invoking GATK tools.
+ */
+public class CountReads {
+  private static final Logger LOG = Logger.getLogger(CountReads.class.getName());
+  private static CountReadsOptions options;
+  private static Pipeline p;
+  private static GenomicsFactory.OfflineAuth auth;
+  public static interface CountReadsOptions extends GenomicsDatasetOptions, GCSOptions {
+    @Description("The ID of the Google Genomics ReadGroupSet this pipeline is working with. "
+        + "Default (empty) indicates all ReadGroupSets.")
+    @Default.String("")
+    String getReadGroupSetId();
+
+    void setReadGroupSetId(String readGroupSetId);
+
+    @Description("The path to the BAM file to get reads data from.")
+    @Default.String("")
+    String getBAMFilePath();
+
+    void setBAMFilePath(String filePath);
+    
+    @Description("Whether to shard BAM reading")
+    @Default.Boolean(true)
+    boolean getShardBAMReading();
+
+    void setShardBAMReading(boolean newValue);
+    
+    class ContigsFactory implements DefaultValueFactory<Iterable<Contig>> {
+      @Override
+      public Iterable<Contig> create(PipelineOptions options) {
+        return Iterables.transform(Splitter.on(",").split(options.as(CountReadsOptions.class).getReferences()),
+            new Function<String, Contig>() {
+              @Override
+              public Contig apply(String contigString) {
+                ArrayList<String> contigInfo = newArrayList(Splitter.on(":").split(contigString));
+                return new Contig(contigInfo.get(0), 
+                    contigInfo.size() > 1 ? 
+                        Long.valueOf(contigInfo.get(1)) : 0, 
+                    contigInfo.size() > 2 ?
+                        Long.valueOf(contigInfo.get(2)) : -1);
+              }
+            });
+      }
+    }
+    
+    @Default.InstanceFactory(ContigsFactory.class)
+    @JsonIgnore
+    Iterable<Contig> getContigs();
```

Comment:
Sounds good.  I'm starting to see a lot of  embedded functions and settings.  Do you think it might help having a collection of functions to try from, until we find the optimal combinations? I get the feeling this code might explode in size :(